src/main/java/org/openhab/tools/analysis/checkstyle/readme/MarkDownVisitor.java

```diff
@@ -0,0 +1,336 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This visitor processes headers, lists and code sections and logs errors when
+ * needed.
+ * 
+ * @author Erdoan Hadzhiyusein - Initial contribution
+ */
+class MarkDownVisitor extends AbstractVisitor {
+
+    private static final String EMPTY_LINE_AFTER_HEADER_MSG = "Missing an empty line after the markdown header ('#').";
+    private static final String EMPTY_LINE_BEFORE_LIST_MSG = "The line before a markdown list must be empty.";
+    private static final String EMPTY_LINE_AFTER_LIST_MSG = "The line after a markdown list must be empty.";
+    private static final String EMPTY_LINE_BEFORE_CODE_MSG = "The line before code formatting section must be empty.";
+    private static final String EMPTY_LINE_AFTER_CODE_MSG = "The line after code formatting section must be empty.";
+    private static final String EMPTY_CODE_BLOCK_WARNING = "There is an empty or unclosed code formatting section. Please correct it.";
+    private final Logger logger = LoggerFactory.getLogger(this.getClass());
+
+    /**
+     * This field stores the line number where the processing of the source is up to.
+     * Every time a MarkDown element is visited the pointer stores its starting line number in the source.
+     * And when it is processed the pointer stores its ending line.
+     */
+    private int currentLinePointer = 0;
+
+    /**
+     * A callback is used in order to use the protected methods of {@link AbstractStaticCheck}
+     */
+    private MarkdownVisitorCallback callBack;
+
+    private File file;
+    private List<String> lines;
+
+    public MarkDownVisitor(MarkdownVisitorCallback callBack, File file, List<String> lines) {
+        this.callBack = callBack;
+        this.file = file;
+        this.lines = lines;
+    }
+
+    /**
+     * Example of heading: #HomeMatic Binding
+     */
+    @Override
+    public void visit(Heading heading) {
+        String headerContent = getLiteralOfElement(heading);
+        validateHeader(headerContent);
+    }
+
+    /**
+     * This method processes the Headings in the README which the MarkDown
+     * parser returns.
+     * 
+     * @param headerValue - a String containing the literal of found Heading node
+     */
+    private void validateHeader(String headerValue) {
+        if (headerValue != null) {
+            int headerLineNumber = 0;
+            headerLineNumber = callBack.findLineNumber(lines, headerValue, currentLinePointer);
+            boolean isInvalidHeaderValue = headerLineNumber == -1;
+            if (isInvalidHeaderValue) {
+                logger.error("A header cannot be processed properly: {}", headerValue);
+            } else {
+                currentLinePointer = headerLineNumber;
+                boolean isHeaderAtEndOfFile = headerLineNumber == lines.size();
+                if (isHeaderAtEndOfFile) {
+                    callBack.log(headerLineNumber, EMPTY_LINE_AFTER_HEADER_MSG);
```

Comment:
As the check only needs to mark missing lines after headers I didn't find any reason to make a specific message for the header position at all. :)